File: Old Java Code/Client/org/hardgate/familytree/filters/LocationFilter.java
/*
 * Created on 02-Jan-2005
 *
 */
package org.hardgate.familytree.filters;

import org.hardgate.familytree.core.Fact;
import org.hardgate.familytree.core.Location;
import org.hardgate.familytree.registrations.Registration;


/**
 * @author db
 *
 */
public class LocationFilter implements RegistrationFilter {

    private String searchString;
    private int level;
    
    public static final LocationFilter SCOTLAND_FILTER =
        	new LocationFilter("Scotland", Location.COUNTRY);
    
    public static final LocationFilter ENGLAND_FILTER =
    	new LocationFilter("England", Location.COUNTRY);

    public LocationFilter (String searchString, int level) {
        this.searchString = searchString.toLowerCase();
        this.level = level; 
    }
    
    public boolean select (Registration r) {
        // If the registration location is not blank and does not
        // contain the search string, then we stop. Otherwise if
        // the registration location is blank, we search all
        // of the facts about this registration.
        Location l = new Location(r.getRegistrationLocation());
        if (!l.isBlank()) {
            return l.matches(searchString, level);
        }
        
        boolean allLocationsBlank = true;
        for (Fact f : r.getAllFacts()) {
            l = new Location(f.getLocation());
            if (l.matches(searchString, level))
                return true;
            if (! l.isBlank())
                allLocationsBlank = false;
        }
        return allLocationsBlank;
    }

}

File: FTAnalyser/Filters/LocationFilter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Xml;

namespace FTAnalyzer
{
    public class LocationFilter : RegistrationFilter {

        private string searchstring;
        private int level;
        
        public static readonly LocationFilter SCOTLAND = new LocationFilter("Scotland", FactLocation.COUNTRY);
        public static readonly LocationFilter ENGLAND = new LocationFilter("England", FactLocation.COUNTRY);
        public static readonly LocationFilter WALES = new LocationFilter("Wales", FactLocation.COUNTRY);
        public static readonly LocationFilter CANADA = new LocationFilter("Canada", FactLocation.COUNTRY);
        public static readonly LocationFilter USA = new LocationFilter("USA", FactLocation.COUNTRY);
        public static readonly LocationFilter US = new LocationFilter("United States", FactLocation.COUNTRY);
        public static readonly LocationFilter IRELAND = new LocationFilter("Ireland", FactLocation.COUNTRY);
        public static readonly LocationFilter EIRE = new LocationFilter("Eire", FactLocation.COUNTRY);

        public LocationFilter(string searchstring, int level)
        {
            this.searchstring = searchstring.ToLower();
            this.level = level; 
        }
        
        public bool select (Registration r) {
            // If the registration location is not blank and does not
            // contain the search string, then we stop. Otherwise if
            // the registration location is blank, we search all
            // of the facts about this registration.
            FactLocation l = new FactLocation(r.RegistrationLocation);
            if (!l.isBlank()) {
                return l.Matches(searchstring, level);
            }
            
            bool allLocationsBlank = true;
            foreach (Fact f in r.AllFacts) {
                l = new FactLocation(f.Place);
                if (l.Matches(searchstring, level))
                    return true;
                if (! l.isBlank())
                    allLocationsBlank = false;
            }
            return allLocationsBlank;
        }
    }
}